### ISSUE TITLE
"KUBE_MAX_PD_VOLS" is ineffective starting from k8s 1.17.

### ISSUE THREAD

**Issue by hardikdr** [kubernetes/kubernetes#92799]
**What happened**: Setting `KUBE_MAX_PD_VOLS ` as an environment variable in kube-scheduler doesn't configure custom-limits for max-attachable volumes. Related doc: https://kubernetes.io/docs/concepts/storage/storage-limits/#custom-limits.

I tried setting `KUBE_MAX_PD_VOLS` to 30 for `c5d.metal` based node in AWS, but kube-scheduler restricts the number of volumes to default (25) for the node. 

This seems to be an issue specifically starting from k8s 1.17+. 
- I have been able to configure the max-attachable volumes limit via env-var with the hack of setting the `AttachVolumeLimit` feature-gate to `false` for k8s 1.16.

**What you expected to happen**: I expected `KUBE_MAX_PD_VOLS`  env-var to overwrite the max-attachable volumes for a kubernetes node. 

**How to reproduce it (as minimally and precisely as possible)**: 
1. Spawn a k8s cluster in AWS with k8s 1.17 or 1.18. 
2. Set the  `KUBE_MAX_PD_VOLS` env-var in kube-scheduler deployment to 3.
3. Try attaching more than 3 persistent-volumes/EBS-disks to any k8s node via statefulset.
4. The scheduler should block the attachment of the 4th volume on a node, but it doesn't.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):  
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-25T15:53:57Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.4", GitCommit:"c96aede7b5205121079932896c4ad89bb93260af", GitTreeState:"clean", BuildDate:"2020-06-17T11:33:59Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: AWS [though the issue is cloud-provider agnostic]

**Comment by hardikdr** [comment]
cc @paurosello

**Comment by liggitt** [comment]
Do you have the exact envvar you're setting? 

> Set the KUBE_MAX_PD__VOLS env-var in kube-scheduler deployment to 3.

`KUBE_MAX_PD__VOLS` looks like a typo (double `__`)

**Comment by hardikdr** [comment]
>KUBE_MAX_PD__VOLS looks like a typo (double __)

Oh yes, that's a typo, I set the exact `KUBE_MAX_PD_VOLS ` as mentioned in the documentation.

**Comment by hardikdr** [comment]
I have opened this, https://github.com/kubernetes/kubernetes/pull/92806, for k8s 1.17. Not sure if that's the best way, it would be nice if someone could take a look @msau42 @bertinatto.

**Comment by bertinatto** [comment]
Hi @hardikdr! 

I took a look at the code and it seems like  the value taken from `KUBE_MAX_PD_VOLS` is overwritten if the `AttachVolumeLimit` feature gate is enabled (it's enabled by default since v1.12). The docs don't make that clear, though, we should probably update them.

As an alternative, you could use the AWS EBS CSI driver as it supports custom volume limits: 

https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/cmd/options/node_options.go#L37

**Comment by hardikdr** [comment]
Unfortunately `AttachVolumeLimit` is disabled/removed permanently since k8s 1.17, that would mean, we can't use the `KUBE_MAX_PD_VOLS`  anymore at all. 

The support for custom-limits was introduced recently in CSI-driver recently by our team, but that doesn't work for a non-csi environment eg k8s 1.17, if not migrated.

**Comment by paurosello** [comment]
We have been able to reproduce what @hardikdr said, not possible to limit PVC on non-CSI environments in k8s 1.17

**Comment by msau42** [comment]
< I tried setting KUBE_MAX_PD_VOLS to 30 for c5d.metal based node in AWS, but kube-scheduler restricts the number of volumes to default (25) for the node.
Usually I see wanting to override the volume limit to reduce the number, not increase. In this case, is the problem that the in-tree driver is not reporting the correct limit for this machine type? cc @bertinatto

**Comment by hardikdr** [comment]
Yes, `c5d.metal` machines in AWS actually support 30 slots for ebs volumes, but k8s defaults it to 25, maybe to reserve few for network-interface as slots are common across both I guess.

**Comment by bertinatto** [comment]
Sorry for the late response, I needed to take some time to look at the code.

The TL;DR is: I don't know if this is the intended behaviour or if it's a bug.

The long story:

There are 2 scheduler plugins (they were predicates in v1.17) that can block pods from being scheduled based on the number of volumes they have: a plugin for CSI volumes and another for non-CSI volumes.

The CSI scheduler plugin counts CSI volumes and in-tree volumes whose plugins have been migrated to CSI drivers. It gets the limit value from (in this order of priority) the CSINode object or from the Node object.

The nonCSI scheduler plugin counts in-tree volumes whose plugins have NOT been migrated to CSI drivers. It gets the limits from (in this order of priority) the Node object, the KUBE_MAX_PD_VOLS env var, or the global per-cloud limits. This scheduler plugin is marked as [deprecated](https://github.com/kubernetes/kubernetes/blob/428b500c5a93daded135f347428724334163859a/pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go#L137).

As state above, the KUBE_MAX_PD_VOLS env var is from the nonCSI scheduler plugin, so this is the one we are interested in.

Now the issue: the precedence of the volume limits value (Node object, env var, cloud) is unchanged since v1.12 [as far as I can see](https://github.com/kubernetes/kubernetes/blob/5f799a487b70aea5e298e5f5f1e3bac904b54ef6/pkg/scheduler/algorithm/predicates/predicates.go#L482-L490) (with the exception that in <v1.17 we could disable the `AttachLimit` feature gate and give the priority to the env var, but that's no longer possible since the gate has graduated to GA).

On the other hand, the docs says that you can change the limits [by setting the env var](https://kubernetes.io/docs/concepts/storage/storage-limits/#custom-limits).

So this is either:

1) An oversight that have been around since v1.12.
2) The intended behaviour && a bug in the docs.

@hekumar, do you have any thoughts on this?

PS: note that the env var should still work with non-cloud storage plugins that don't specify the limit in the Node object.

**Comment by msau42** [comment]
@bertinatto I was asking if the logic in https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/awsebs/aws_ebs.go#L102 needs to be updated to account for this machine type?

**Comment by bertinatto** [comment]
> @bertinatto I was asking if the logic in https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/awsebs/aws_ebs.go#L102 needs to be updated to account for this machine type?

Sorry, I missed your question.

IMO, yes - this value (25) is quite generic and applied to [a few instance types based on the Nitro Sistem](https://github.com/kubernetes/kubernetes/blob/1fdd8fb213e0361e8f18b1dd152dddb4c88ad183/pkg/volume/util/attach_limit.go#L31).

However,  AWS docs [aren't very specific about the limits for each instance type](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html), plus these limits are shared with network interface attachments. This makes it difficult to choose a "correct" value. I believe this is the reason this code is this generic.

CC @leakingtapan @wongma7

**Comment by msau42** [comment]
If we can get a more accurate dynamic limit, that would be ideal. The problem with setting KUBE_MAX_PD_VOLS is it applies to all nodes in the cluster, so if your cluster has machines of different sizes with different limits, you'll have to use the min value that works across all.

**Comment by paurosello** [comment]
> However, AWS docs [aren't very specific about the limits for each instance type](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html), plus these limits are shared with network interface attachments. This makes it difficult to choose a "correct" value. I believe this is the reason this code is this generic.

The fact that attachments are shared with network interfaces is the reason we have to use this feature, with AWS CNI pods will be using additional interfaces and the limit will be off
